cpp/src/arrow/python/decimal.cc

```diff
@@ -145,8 +146,9 @@ Status DecimalFromPythonDecimal(PyObject* python_decimal, const DecimalType& arr
   return DecimalFromStdString(string, arrow_type, out);
 }
 
-Status DecimalFromPyObject(PyObject* obj, const DecimalType& arrow_type,
-                           Decimal128* out) {
+template <typename ArrowDecimal>
+Status InternalDecimalFromPyObject(PyObject* obj, const DecimalType& arrow_type,
+                                   ArrowDecimal* out) {
```

Comment:
In the implementation below you should call `InternalDecimalFromPythonDecimal<ArrowDecimal>` instead of `DecimalFromPythonDecimal`.

Reply:
done.